File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/SettingsPositionsActivity.java
package com.saienko.androidthings.barman.ui.settings;

import android.content.Intent;
import android.os.Bundle;
import com.saienko.androidthings.barman.R;
import com.saienko.androidthings.barman.db.AppDatabase;
import com.saienko.androidthings.barman.db.DatabaseUtil;
import com.saienko.androidthings.barman.db.position.Position;
import com.saienko.androidthings.barman.ui.base.BaseListActivity;
import com.saienko.androidthings.barman.ui.dialog.CustomDialog;
import com.saienko.androidthings.barman.ui.settings.adapter.PositionAdapter;
import io.reactivex.Completable;
import io.reactivex.Single;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.schedulers.Schedulers;

public class SettingsPositionsActivity extends BaseListActivity {

    private static final int REG_CODE_ADD_POSITION  = 32;
    private static final int REG_CODE_EDIT_POSITION = 33;
    private PositionAdapter adapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_settings_positions);
        initUI();
        getPositionList();
    }


    @Override
    public void initUI() {
        super.initUI();
        initRecyclerView(findViewById(R.id.list));
    }


    private void getPositionList() {
        Single.fromCallable(DatabaseUtil::getPositions).subscribeOn(Schedulers.io())
              .observeOn(AndroidSchedulers.mainThread())
              .subscribe(positions -> getAdapter().addItems(positions), this::showErrorSnackBar);
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (resultCode == RESULT_OK) {
            if (requestCode == REG_CODE_ADD_POSITION) {
                Position position = data.getParcelableExtra(AddPositionActivity.EXTRA_POSITION);
                savePosition(position);
            }
            if (requestCode == REG_CODE_EDIT_POSITION) {
                Position position = data.getParcelableExtra(AddPositionActivity.EXTRA_POSITION);
                updatePosition(position);
            }
        }
    }

    private void savePosition(Position position) {
        Completable.fromCallable(() -> {
            // TODO: 9/27/17 adb shell monkey UNIQUE constraint failed: Position.id (code 1555)
            AppDatabase.getDb().positionDao().insert(position);
            getPositionList();
            return null;
        }).subscribeOn(Schedulers.io())
                   .observeOn(AndroidSchedulers.mainThread())
                   .subscribe();
    }

    private void updatePosition(Position position) {
        Completable.fromCallable(() -> {
            AppDatabase.getDb().positionDao().update(position);
            getPositionList();
            return null;
        }).subscribeOn(Schedulers.io())
                   .observeOn(AndroidSchedulers.mainThread())
                   .subscribe();
    }

    @Override
    protected void removeAll() {
        Completable
                .fromCallable(() -> {
                    AppDatabase.getDb().positionDao().clear();
                    getPositionList();
                    return null;
                })
                .subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }

    @Override
    protected PositionAdapter getAdapter() {
        if (adapter == null) {
            adapter = new PositionAdapter(new PositionAdapter.OnItemListener() {
                @Override
                public void onDelete(Position position) {
                    showDeleteDialog(position);
                }

                @Override
                public void onItemClick(Position position) {
                    editPosition(position);
                }
            });
        }
        return adapter;
    }

    private void editPosition(Position position) {
        if (position.getComponent() != null) {
            AddPositionActivity.edit(SettingsPositionsActivity.this, position, REG_CODE_EDIT_POSITION);
        } else {
            AddPositionActivity.add(SettingsPositionsActivity.this, position, REG_CODE_ADD_POSITION);
        }
    }

    private void showDeleteDialog(Position position) {
        CustomDialog.showDialog(this,
                                "Are you sure ?",
                                "Do you want to delete this item ?",
                                (dialogInterface, i) -> deleteItem(position));
    }

    private void deleteItem(Position position) {
        Completable
                .fromCallable(() -> {
                    AppDatabase.getDb().positionDao().delete(position);
                    getPositionList();
                    return null;
                }).subscribeOn(Schedulers.io())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe();
    }
}

File: app/src/main/java/com/saienko/androidthings/barman/ui/settings/SettingsPositionsActivity.kt
package com.saienko.androidthings.barman.ui.settings

import android.content.DialogInterface
import android.os.Bundle
import com.saienko.androidthings.barman.R
import com.saienko.androidthings.barman.db.entity.Position
import com.saienko.androidthings.barman.db.interfaces.IBaseManager
import com.saienko.androidthings.barman.db.manager.PositionManager
import com.saienko.androidthings.barman.ui.base.BaseAdapter
import com.saienko.androidthings.barman.ui.base.BaseListActivity
import com.saienko.androidthings.barman.ui.dialog.CustomDialog
import com.saienko.androidthings.barman.ui.settings.adapter.PositionAdapter
import com.saienko.androidthings.barman.ui.settings.adapter.holder.PositionViewHolder

class SettingsPositionsActivity : BaseListActivity<Position, PositionViewHolder>() {
    override val adapter: BaseAdapter<Position, PositionViewHolder>
        get() = localAdapter
    override val baseManager: IBaseManager<Position>
        get() = PositionManager()

    private lateinit var localAdapter: PositionAdapter

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_settings_positions)
        initUI()
    }

    override fun initUI() {
        super.initUI()
        initRecyclerView(findViewById(R.id.list))
    }

    override fun initAdapter() {
        localAdapter = PositionAdapter(object : PositionAdapter.OnItemListener {
            override fun onDelete(position: Position) {
                showDeleteDialog(position)
            }

            override fun onItemClick(position: Position) {
                editPosition(position)
            }
        })
    }

    private fun editPosition(position: Position) {
        if (position.component != null) {
            AddPositionActivity.edit(this@SettingsPositionsActivity, position.id)
        } else {
            AddPositionActivity.add(this@SettingsPositionsActivity, position.id, position.motorId)
        }
    }

    private fun showDeleteDialog(position: Position) {
        CustomDialog.showDialog(this,
                                getString(R.string.are_you_sure),
                                getString(R.string.do_you_want_to_delete_this_item),
                                DialogInterface.OnClickListener { _, _ -> deleteItem(position) })
    }
}
